### ISSUE TITLE
Raspbian Stretch + VC4 + Qt5 = needs some love

### ISSUE THREAD

**Issue by mcallegari** [RPi-Distro/repo#105]
Hi, instead of commenting on a closed discussion (#16) I'd like to start a new one, because the situation of Qt5 on Raspbian Stretch is pretty messed up.

I built many Qt versions on Jessie and never had an issue.
Please let me try to explain what I've seen on Stretch in the last couple of weeks.

- the provided 5.7.1 is broken on Xorg. If I launch my application, there are no icons at all. Tried all the possible flags from this bug report (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=803423) and none worked.
- the decision to rename /opt/vc/lib libraries by prefixing them with 'brcm' IMHO is a poor one. It broke many existing softwares, including Qt. See here: http://code.qt.io/cgit/qt/qtbase.git/tree/mkspecs/devices/linux-rasp-pi3-g++/qmake.conf
- to make even more mess, keeping Mesa VC4 together with Broadcom VC results in duplicated headers files in /usr/include and in /opt/vc/include. Files with the same name, but different code (e.g. eglplatform.h). And Qt refuses to build.
- building Qt 5.10.1 on the Raspberry itself, seems to be valid, but it's not. Launched my application on Xorg (with -platform xcb)  -> segfault. Same application works fine with -platform eglfs.
- for booting time reasons, my usage case has always been EGLFS on proprietary BCM VC libraries, without Xorg. My application UI is displayed 12 seconds after powering up. This is still possible BUT you need to stay away from any Xorg development package. While on Jessie I could build Qt just once and have EGLFS **and** XCB QPA plugins, on Stretch this is no longer possible. When Qt detects Mesa, the /opt/vc libraries are ignored, therefore breaking my usage case. Tried to report to the Qt bugtracking, closed right away: https://bugreports.qt.io/browse/QTBUG-66465 and https://bugreports.qt.io/browse/QTBUG-62216
- correct me if I'm wrong, but Mesa depends on Xorg. If in the long run the proprietary BCM libraries are set to disappear, how will I be able to render on a OpenGL surface like I do now on EGLFS without Xorg ? Maybe @anholt can comment on this ? Cause there'd be no feature parity with the excellent work he's doing.

In conclusion, this is a mess. I am not able to provide to my users a solution that works with **and** without Xorg as it was with Jessie.

IMHO BCM VC and Mesa VC4 should never exist together. Maybe it could be simple to address by tweaking raspi-config ?

I'd appreciate some help, cause I'm in a position where it seems to be impossible to resolve this situation on my own.
Thanks

**Comment by XECDesign** [comment]
I'm not a Qt expert, so I'm not exactly sure what you're saying it broken. There are a ton of Qt applications included in Raspbian and seem to work well. If there's something a bit more specialised that's not needed by 99% of users, support can be hard to come by.

We're moving away from the proprietary libraries, so Mesa should be used instead whenever possible.

Yes, renaming the libraries isn't pretty and broke existing software but of all the options discussed it was the least worst. The issue with conflicting header files is definitely one that needs some thought. In general, don't use /opt/vc/includes if you don't need them, but that can be tricky if you need some of the non-GL ones.

A lot of this seems to be something that should be addressed upstream by Qt or Debian. Raspbian, for the most part, just pulls in what's available in Debian.

Let me know if this is okay to close or if there's anything we can do. If you have a patch that preserves all existing functionality and extends it further, I'd have no problem with adding our own patches version of Qt.
